modules/standard/Random.chpl

```diff
@@ -1188,55 +1182,48 @@ module Random {
                                  seed, PCGRandomStreamPrivate_count,
                                  0, i);
 
-          var j = i;
-
-          // Strided case
-          if stride > 1 {
-            k *= stride;
-            j *= stride;
-          }
-
-          // Alignment offsets
-          k += low;
-          j += low;
-
-          arr[k] <=> arr[j];
+          var lastInd = D.orderToIndex(i);
+          var randInd = D.orderToIndex(k);
+          arr[lastInd] <=> arr[randInd];
         }
 
         PCGRandomStreamPrivate_count += D.size;
 
         _unlock();
       }
 
-      /* Produce a random permutation, storing it in a 1-D array.
-         The resulting array will include each value from low..high
-         exactly once, where low and high refer to the array's domain.
+      /* Produce a random permutation, storing it in an array.
+         The resulting array will include each value from 1..n
+         exactly once, where n is the array size.
          */
       proc permutation(arr: [] eltType) {
 
         if(!isRectangularArr(arr)) then
           compilerError("permutation does not support non-rectangular arrays");
-
-        var low = arr.domain.dim(0).low;
-        var high = arr.domain.dim(0).high;
-
-        if arr.domain.rank != 1 then
-          compilerError("Permutation requires 1-D array");
-        //if arr.domain.dim(0).stridable then
-        //  compilerError("Permutation requires non-stridable 1-D array");
+        
+        if(!isNumericType(arr)) then
+          compilerError("permutation does not support non-numeric arrays");
 
         _lock();
 
-        for i in low..high {
+        for i in 0..(arr.size-1) {
+          var ind = arr.domain.orderToIndex(i);
+          arr[ind] = (i+1);
+        }
+
+        for i in 0..(arr.size-1) by -1 {
```

Comment:
This can also be:

```chpl
for i in 0..<arr.size by -1 {
```

Why are we looping in reverse order now?

Reply:
I think that the previous version of Fisher Yates shuffle used for permuting array indices previously was incorrect. The version of shuffle I'm introducing here is simply the one we're using in procedure shuffle()